### PR TITLE
C51-386: Setup initial backstop test suite for civicase 

### PR DESCRIPTION
**Pull request by swastikpareek**
## Overview
This PR setup basic backstop test suite and adds gulp tasks related to it with proper documentations

## Description
Backstop test suite which is based on visual regression testing is being added inside `test` folder. The new gulp tasks are available to create reference screenshots and tests that reference screenshots after working on the designs
As a part of this PR, following tasks are made available
```shell
$ gulp backstopjs:reference
$ gulp backstopjs:test
$ gulp backstopjs:approve
$ gulp backstopjs:openReport
```
Civicase Dashboard screen is also covered as a part of this PR
civicase_-_dashboard

## Technical Details
The PR takes code excerpt from https://github.com/civicrm/org.civicrm.civicase/pull/112/ and contains predefined bootstrapping and some utility files.

Now while taking the screenshot on dashboard screen the challenge was to 
1. wait for the angular to load the app on the page and then 
2. wait for all content/requests to load

The *1* problem was solved by waiting for `#bootstrap-element` to appear on the screen.
```js
await engine.waitForSelector('#bootstrap-theme');
```

After waiting for it the challenge *2* was solved by waiting for all placeholder divs to vanish from the views

```js
   await engine.waitFor(() => {
      const allLoadingElements = document.querySelectorAll('div[class*="civicase__loading-placeholder"]');

      return allLoadingElements.length === 0;
    });
```

### CODE REVIEW

File: gulpfile.js [hunk 1]
```diff
@@ -10,25 +10,208 @@
  */
 'use strict';
 
-var gulp = require('gulp');
+var _ = require('lodash');
```
Comment (deb1990) [inline gulpfile.js]: The file has become too long, we need to look into breaking this. Lets discuss this in a call.

File: tests/backstop_data/engine_scripts/puppet/clickAndHoverHelper.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
+module.exports = async (page, scenario) => {
```
[the remark this answers is not in the document]
Comment (swastikpareek) [reply tests/backstop_data/engine_scripts/puppet/clickAndHoverHelper.js]: Cool make sense. Let me remove it

File: tests/backstop_data/engine_scripts/puppet/ignoreCSP.js [hunk 3]
```diff
@@ -0,0 +1,65 @@
+/**
```
[the remark this answers is not in the document]
Comment (swastikpareek) [reply tests/backstop_data/engine_scripts/puppet/ignoreCSP.js]: I think these files got added while setting up backstop test suite using `backstop init` command. These are not necessary for the current project. So removing it

File: tests/backstop_data/engine_scripts/puppet/interceptImages.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
+/**
```
[the remark this answers is not in the document]
Comment (swastikpareek) [reply tests/backstop_data/engine_scripts/puppet/interceptImages.js]: I think these files got added while setting up backstop test suite using `backstop init` command. These are not necessary for the current project. So removing it

File: gulpfile.js (hunk 1 above)
Comment (swastikpareek) [reply gulpfile.js]: Good idea. I have created a new export file for backstopJS task. Now the file is smaller.

File: TESTING.md [hunk 5]
```diff
@@ -64,7 +64,7 @@ This test suite is based on [BackstopJS](https://garris.github.io/BackstopJS) pl
 Documentation available [here](https://github.com/garris/BackstopJS#backstopjs)
 
 ### Requirements
-A CiviCRM with pre-configured with CiviCRM, Shoreditch, CiviCase, and any other dependencies. [See how](https://github.com/compucorp/uk.co.compucorp.civicase#installation-civibuild)
+A CiviCRM with pre-configured CiviCRM, Shoreditch, CiviCase, and any other dependencies. [See how](https://github.com/compucorp/uk.co.compucorp.civicase#installation-civibuild)
```
Comment (deb1990) [inline TESTING.md]: I am sorry but this still does not sound right
`A CiviCRM with pre-configured CiviCRM, Shoreditch, CiviCase, and any other dependencies`

notice the CiviCRM is present twice.

File: gulpfile.js [hunk 6]
```diff
@@ -10,21 +10,22 @@
...
+var postcssPrefix = require('postcss-prefix-selector');
 var rename = require('gulp-rename');
+var sass = require('gulp-sass');
+var stripCssComments = require('gulp-strip-css-comments');
 var sourcemaps = require('gulp-sourcemaps');
-var autoprefixer = require('gulp-autoprefixer');
-var karma = require('karma');
-var path = require('path');
+var transformSelectors = require('gulp-transform-selectors');
+var runBackstopJS = require('./gulp-tasks/backstopjs.js');
```
Comment (deb1990) [inline gulpfile.js]: Lets call this, something like `backstopjsTask`. Then we can follow `<>-task` syntax in future.

File: gulpfile.js [hunk 7]
```diff
@@ -100,3 +101,15 @@ gulp.task('test', function (done) {
...
 });
+
+/**
+  * Gulp backstop tasks
+  * 'backstopjs:reference': For creating reference screenshots
+  * 'backstopjs:test': For creating test screenshots and matching them
+  * 'backstopjs:openReport': For opening reports in the browser
+  * 'backstopjs:approve': Approving reports
+  */
+
```
Comment (deb1990) [inline gulpfile.js]: Remove extra line gap.

**Pull request merged**